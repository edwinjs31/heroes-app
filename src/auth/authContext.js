import { createContext } from "react";


//contexto para proveer informacion a los componentes hijos, retorna un Componente
//Ã©ste componente debe ser implementado en el componente mas alto
export const AuthContext = createContext();