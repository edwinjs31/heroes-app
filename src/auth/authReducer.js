import { types } from "../types/types";

//LOS REDUCER SON FUNCIONES PURAS

//el 'action' es la accion que va modificar el state
export const authReducer = (state = {}, action) => {
    switch (action.type) {
        case types.login:
            return {
                ...action.payload,
                logged: true
            }
        case types.logout:
            return {
                logged: false
            }

        default:
            return state;
    }
}
