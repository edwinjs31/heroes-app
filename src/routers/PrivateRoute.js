import { useContext } from "react"
import { Navigate, useLocation } from "react-router-dom";
import { AuthContext } from "../auth/authContext";

export const PrivateRoute = ({ children }) => {

    const { user } = useContext(AuthContext);
    const {pathname, search} = useLocation();

    //guardamos el ultimo path, al volver a loguearse iniciará desde ese ultimo punto.
    //search es un queryString
    localStorage.setItem('lastPath', pathname + search);


    //si está logueado devuelve todas las rutas privadas, si no redirecciona al login
    return user.logged ? children : <Navigate to='/login' />

}
