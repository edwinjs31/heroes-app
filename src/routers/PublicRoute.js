import { useContext } from "react"
import { Navigate } from "react-router-dom";
import { AuthContext } from "../auth/authContext"

export const PublicRoute = ({ children }) => {

    const { user } = useContext(AuthContext);

    //si est√° loquedado redirecciona al home, si no al login
    return user.logged ? <Navigate to='/' /> : children;

}
