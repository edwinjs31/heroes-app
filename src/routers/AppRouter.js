//1º ROUTER PRINCIPAL MANEJARÁ LAS RUTAS ANTES DE HACER LOGIN
import React from 'react'
import { BrowserRouter, Route, Routes } from 'react-router-dom'
import { LoginScreen } from '../components/login/LoginScreen'
import { DashBoardRoutes } from './DashBoardRoutes'
import { PrivateRoute } from './PrivateRoute'
import { PublicRoute } from './PublicRoute'


export const AppRouter = () => {
    return (
        <BrowserRouter>
            <Routes>
                <Route path="/login" element={
                    <PublicRoute>
                        <LoginScreen />
                    </PublicRoute>
                } />

                {/* <Route path="/login" element={<LoginScreen />} /> */}

                {/* Despues del login todas las rutas serán manejadas PrivateRoute, quien validará si está logueado */}
                <Route path="/*" element={
                    <PrivateRoute>
                        <DashBoardRoutes /> {/*Estas rutas hijas serán enviadas como 'props' a 'PrivateRoute'*/}
                    </PrivateRoute>
                } />

                {/* <Route path="/*" element={<DashBoardRoutes />} /> */}
            </Routes>
        </BrowserRouter>

    )
}
