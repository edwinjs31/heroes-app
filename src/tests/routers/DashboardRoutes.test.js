import { DashBoardRoutes } from "../../routers/DashBoardRoutes";
import { mount } from 'enzyme';
import { AuthContext } from "../../auth/authContext";
import { MemoryRouter } from "react-router-dom";

describe('Pruebas en DashboardRoutes', () => {
    const contexValue = {
        user: {
            logged: true,
            name: 'Kevin'
        }
    }

    test('Debe de mostrarse correctamente', () => {
        const wrapper = mount(
            <AuthContext.Provider value={contexValue}>
                <MemoryRouter>
                    <DashBoardRoutes />
                </MemoryRouter>

            </AuthContext.Provider>
        );
        //expect(wrapper).toMatchSnapshot();
        expect(wrapper.find('.text-info').text().trim()).toBe('Kevin');
    })
})