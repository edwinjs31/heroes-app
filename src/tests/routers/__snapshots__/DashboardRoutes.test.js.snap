// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en DashboardRoutes Debe de mostrarse correctamente 1`] = `
Array [
  <nav
    className="navbar navbar-expand-md navbar-dark bg-dark"
  >
    <div
      className="container-fluid"
    >
      <a
        className="navbar-brand ms-4"
        href="/"
        onClick={[Function]}
      >
        Asociaciones
      </a>
      <button
        aria-controls="minavBar"
        aria-expanded="false"
        aria-label="Toggle navigation"
        className="navbar-toggler"
        data-bs-target="#minavBar"
        data-bs-toggle="collapse"
        type="button"
      >
        <span
          className="navbar-toggler-icon"
        />
      </button>
      <div
        className="collapse navbar-collapse"
        id="minavBar"
      >
        <div
          className="navbar-nav"
        >
          <a
            className="nav-item nav-link"
            href="/marvel"
            onClick={[Function]}
          >
            Marvel
          </a>
          <a
            className="nav-item nav-link"
            href="/dc"
            onClick={[Function]}
          >
            DC
          </a>
          <a
            className="nav-item nav-link"
            href="/search"
            onClick={[Function]}
          >
            Search
          </a>
        </div>
        <div
          className="w-100 d-flex justify-content-end me-4"
        >
          <ul
            className="navbar-nav "
          >
            <span
              className="nav-item nav-link text-info"
            />
            <button
              className="nav-item nav-link btn btn-outline-info"
              onClick={[Function]}
            >
              Logout
            </button>
          </ul>
        </div>
      </div>
    </div>
  </nav>,
  <div
    className="container"
  >
    <div>
      <h1>
        Heroes Marvel Comics
      </h1>
      <hr />
      <div
        className="row rows-cols-1 row-cols-md-2 row-cols-lg-3 animate__animated animate__pulse"
      >
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Spider Man"
                  className="card-img"
                  src="/assets/img/marvel-spider.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Spider Man
                  </h5>
                  <p
                    className="card-text"
                  >
                    Peter Parker
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Amazing Fantasy #15
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-spider"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Captain America"
                  className="card-img"
                  src="/assets/img/marvel-captain.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Captain America
                  </h5>
                  <p
                    className="card-text"
                  >
                    Steve Rogers
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Captain America Comics #1
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-captain"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Iron Man"
                  className="card-img"
                  src="/assets/img/marvel-iron.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Iron Man
                  </h5>
                  <p
                    className="card-text"
                  >
                    Tony Stark
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Tales of Suspense #39
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-iron"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Thor"
                  className="card-img"
                  src="/assets/img/marvel-thor.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Thor
                  </h5>
                  <p
                    className="card-text"
                  >
                    Thor Odinson
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Journey into Myster #83
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-thor"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Hulk"
                  className="card-img"
                  src="/assets/img/marvel-hulk.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Hulk
                  </h5>
                  <p
                    className="card-text"
                  >
                    Bruce Banner
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      The Incredible Hulk #1
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-hulk"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Wolverine"
                  className="card-img"
                  src="/assets/img/marvel-wolverine.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Wolverine
                  </h5>
                  <p
                    className="card-text"
                  >
                    James Howlett
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      The Incredible Hulk #180
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-wolverine"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Daredevil"
                  className="card-img"
                  src="/assets/img/marvel-daredevil.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Daredevil
                  </h5>
                  <p
                    className="card-text"
                  >
                    Matthew Michael Murdock
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Daredevil #1
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-daredevil"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Hawkeye"
                  className="card-img"
                  src="/assets/img/marvel-hawkeye.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Hawkeye
                  </h5>
                  <p
                    className="card-text"
                  >
                    Clinton Francis Barton
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      Tales of Suspense #57
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-hawkeye"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Cyclops"
                  className="card-img"
                  src="/assets/img/marvel-cyclops.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Cyclops
                  </h5>
                  <p
                    className="card-text"
                  >
                    Scott Summers
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      X-Men #1
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-cyclops"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          className="row g-2 "
        >
          <div
            className="card"
          >
            <div
              className="row no-gutters"
            >
              <div
                className="col-4 animate__animated animate__rotateInUpLeft"
              >
                <img
                  alt="Silver Surfer"
                  className="card-img"
                  src="/assets/img/marvel-silver.jpg"
                />
              </div>
              <div
                className="col-8"
              >
                <div
                  className="car-body"
                >
                  <h5
                    className="card-title"
                  >
                    Silver Surfer
                  </h5>
                  <p
                    className="card-text"
                  >
                    Norrin Radd
                  </p>
                  <p
                    className="card-text"
                  >
                    <small
                      className="text-muted"
                    >
                      The Fantastic Four #48
                    </small>
                  </p>
                  <a
                    href="/hero/marvel-silver"
                    onClick={[Function]}
                  >
                    Más...
                  </a>
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
]
`;
